src/http_client_async.zig

```diff
@@ -51,6 +51,144 @@ var shared_response_headers_buf: [256]picohttp.Header = undefined;
+            bun.default_allocator.free(data);
+        }
+    }
+};
+
+const ProxyTunnel = struct {
+    ssl_ctx: *BoringSSL.SSL_CTX,
+    ssl: *BoringSSL.SSL,
+    out_bio: *BoringSSL.BIO,
+    in_bio: *BoringSSL.BIO,
+    partial_data: ?ProxySSLData,
+    read_buffer: []u8,
+
+    pub fn init(comptime is_ssl: bool, client: *HTTPClient, socket: NewHTTPContext(is_ssl).HTTPSocket) ProxyTunnel {
+        BoringSSL.load();
+        var context = BoringSSL.SSL_CTX.init();
+
+        if ( context ) | ssl_context | {
+            var ssl_ctx = ssl_context;
+            var ssl = BoringSSL.SSL.init(ssl_context);
+            ssl.setIsClient(true);            
+            var out_bio: *BoringSSL.BIO = undefined;
+            if (comptime is_ssl){    
+                //TLS -> TLS
+                var proxy_ssl: *BoringSSL.SSL = @ptrCast(*BoringSSL.SSL, socket.getNativeHandle());
+                //create new SSL BIO
+                out_bio = BoringSSL.BIO_new(BoringSSL.BIO_f_ssl()) orelse unreachable;
+                //chain SSL bio with proxy BIO
+                var proxy_bio = BoringSSL.SSL_get_wbio(proxy_ssl);
+                _ = BoringSSL.BIO_push(out_bio, proxy_bio);
+
+            }else {
+                // socket output bio for non-TLS -> TLS
+                var fd = @intCast(c_int, @ptrToInt(socket.getNativeHandle()));
+                out_bio = BoringSSL.BIO_new_fd(fd, BoringSSL.BIO_NOCLOSE);
+            }
+            
+            // in memory bio to control input flow from onData handler
+            var in_bio = BoringSSL.BIO.init() catch {
+                unreachable;
+            };
+            _ = BoringSSL.BIO_set_mem_eof_return(in_bio, -1);
+            ssl.setBIO(in_bio, out_bio);
+
+            const hostname = bun.default_allocator.dupeZ(u8, client.url.hostname) catch unreachable;
+            defer bun.default_allocator.free(hostname);
+
+            ssl.configureHTTPClient(hostname);
+            BoringSSL.SSL_CTX_set_verify(ssl_ctx, BoringSSL.SSL_VERIFY_NONE, null);
+            BoringSSL.SSL_set_verify(ssl, BoringSSL.SSL_VERIFY_NONE, null);
+            return ProxyTunnel {
+                .ssl = ssl,
+                .ssl_ctx = ssl_ctx,
+                .in_bio = in_bio,
+                .out_bio = out_bio,
+                .read_buffer = bun.default_allocator.alloc(u8, 16 * 1024) catch unreachable,
+                .partial_data = null
+            };
+        }
+        unreachable;
+    }
+    
+    pub fn getSSLData(this: *@This(), incoming_data: ?[]const u8) !ProxySSLData {
+         
+         if (incoming_data) | data | {
+             _ = this.in_bio.write(data) catch  {
+                return error.OutOfMemory;
+            };
+         }
+        
+        var data: ProxySSLData = undefined;
+        if (this.partial_data) | partial | {
+            data = partial;
+            data.partial = false;
+        } else {
+            data = ProxySSLData.init() catch |err| {
```

Comment:
`catch |err| return err` is equivalent to `try {expression}`

Reply:
done